File: backend/src/application/agent/ProductOptimizationPipeline.ts
import { AgentRun } from "../../domain/entities/AgentRun.js";
import { ProductListing } from "../../domain/entities/ProductListing.js";
import { AgentRunRepository } from "../../domain/repositories/AgentRunRepository.js";
import { AgentStepRunRepository } from "../../domain/repositories/AgentStepRunRepository.js";
import { OptimizationSuggestionRepository } from "../../domain/repositories/OptimizationSuggestionRepository.js";
import { OptimizationDraft } from "./OptimizationDraft.js";
import { ListingAnalyzer } from "./StructuredListingAnalyzer.js";
import { AgentStepName } from "./OptimizationProvider.js";
import { PromptBuilder } from "./OptimizationPromptBuilder.js";
import { OptimizationSuggestionGenerator } from "./OptimizationSuggestionGenerator.js";
import { ProductSnapshotTool } from "./ProductSnapshot.js";

const pipelineSteps: AgentStepName[] = [
  "listing_snapshot",
  "seo_analysis",
  "copy_analysis",
  "tag_analysis",
  "audience_positioning_analysis",
  "final_synthesis"
];

export class ProductOptimizationPipeline {
  constructor(
    private readonly snapshotTool: ProductSnapshotTool,
    private readonly analyzer: ListingAnalyzer,
    private readonly promptBuilder: PromptBuilder,
    private readonly generator: OptimizationSuggestionGenerator,
    private readonly agentRuns: AgentRunRepository,
    private readonly agentStepRuns: AgentStepRunRepository,
    private readonly suggestions: OptimizationSuggestionRepository
  ) {}

  async run(product: ProductListing): Promise<AgentRun> {
    const agentRun = await this.agentRuns.createRunning(product.id, { pipeline: "product-listing-optimization" });

    try {
      const snapshot = this.snapshotTool.execute(product);
      const analysis = this.analyzer.analyze(snapshot);
      const stepResponses: Record<string, string> = {};
      let finalDraft: OptimizationDraft | null = null;

      for (const [index, stepName] of pipelineSteps.entries()) {
        const prompt = this.promptBuilder.buildStep(stepName, snapshot, analysis, stepResponses);
        const stepRun = await this.agentStepRuns.createRunning({
          agentRunId: agentRun.id,
          stepName,
          stepOrder: index + 1,
          prompt
        });

        try {
          const result = await this.generator.generate(stepName, prompt, snapshot, analysis, stepResponses);
          await this.agentStepRuns.markCompleted(stepRun.id, {
            response: result.response,
            provider: result.provider,
            model: result.model,
            inputTokens: result.inputTokens,
            outputTokens: result.outputTokens,
            estimatedCost: result.estimatedCost
          });
          stepResponses[stepName] = result.response;

          if (stepName === "final_synthesis") {
            finalDraft = result.draft ?? null;
          }
        } catch (error) {
          const message = error instanceof Error ? error.message : "Unknown step failure";
          await this.agentStepRuns.markFailed(stepRun.id, message);
          throw error;
        }
      }

      if (!finalDraft) {
        throw new Error("Final synthesis did not produce a valid optimization suggestion.");
      }

      await this.suggestions.create({
        productListingId: product.id,
        agentRunId: agentRun.id,
        ...finalDraft
      });

      return this.agentRuns.markCompleted(agentRun.id, {
        ...analysis,
        stepFindings: stepResponses
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown pipeline failure";
      return this.agentRuns.markFailed(agentRun.id, message);
    }
  }
}

File: app/services/agent/product_optimization_pipeline.rb
module Agent
  class ProductOptimizationPipeline
    PIPELINE_STEPS = STEP_NAMES.freeze

    def initialize(
      snapshot_tool: ProductSnapshotTool.new,
      analyzer: StructuredListingAnalyzer.new,
      prompt_builder: OptimizationPromptBuilder.new,
      generator: OptimizationSuggestionGenerator.new
    )
      @snapshot_tool  = snapshot_tool
      @analyzer       = analyzer
      @prompt_builder = prompt_builder
      @generator      = generator
    end

    def run(listing)
      agent_run = AgentRun.create!(
        product_listing: listing,
        status: "running",
        started_at: Time.current,
        metadata: { pipeline: "product-listing-optimization" }
      )

      snapshot       = @snapshot_tool.execute(listing)
      analysis       = @analyzer.analyze(snapshot)
      prior_responses = {}
      final_draft    = nil

      PIPELINE_STEPS.each_with_index do |step_name, index|
        prompt   = @prompt_builder.build_step(step_name, snapshot, analysis, prior_responses)
        step_run = AgentStepRun.create!(
          agent_run: agent_run,
          step_name: step_name,
          step_order: index + 1,
          status: "running",
          prompt: prompt
        )

        begin
          result = @generator.generate(
            step_name: step_name,
            prompt: prompt,
            snapshot: snapshot,
            analysis: analysis,
            prior_responses: prior_responses
          )

          step_run.update!(
            status: "completed",
            response: result[:response],
            provider: result[:provider],
            model: result[:model],
            input_tokens: result[:input_tokens],
            output_tokens: result[:output_tokens],
            estimated_cost: result[:estimated_cost]
          )

          prior_responses[step_name] = result[:response]
          final_draft = result[:draft] if step_name == "final_synthesis"
        rescue => e
          step_run.update!(status: "failed", error_message: e.message.truncate(1000))
          raise
        end
      end

      raise "Final synthesis did not produce a valid optimization suggestion." unless final_draft

      OptimizationSuggestion.create!(
        product_listing: listing,
        agent_run: agent_run,
        improved_title: final_draft[:improved_title],
        improved_description: final_draft[:improved_description],
        tags: final_draft[:tags],
        positioning_recommendations: final_draft[:positioning_recommendations],
        seo_notes: final_draft[:seo_notes],
        score: final_draft[:score],
        rationale: final_draft[:rationale]
      )

      agent_run.update!(
        status: "completed",
        completed_at: Time.current,
        analysis: analysis.merge(step_findings: prior_responses)
      )
      agent_run.reload
    rescue => e
      agent_run&.update!(
        status: "failed",
        completed_at: Time.current,
        error_message: e.message.truncate(1000)
      )
      agent_run&.reload
    end
  end
end
